src/DynamicData/Cache/Internal/ExpireAfter.ForSource.cs

```diff
@@ -0,0 +1,399 @@
+        where TKey : notnull
+    {
+        public static IObservable<IEnumerable<KeyValuePair<TKey, TObject>>> Create(
+            ISourceCache<TObject, TKey> source,
+            Func<TObject, TimeSpan?> timeSelector,
+            TimeSpan? pollingInterval = null,
+            IScheduler? scheduler = null)
+        {
+            source.ThrowArgumentNullExceptionIfNull(nameof(source));
+            timeSelector.ThrowArgumentNullExceptionIfNull(nameof(timeSelector));
+
+            return Observable.Create<IEnumerable<KeyValuePair<TKey, TObject>>>(observer => (pollingInterval is TimeSpan pollingIntervalValue)
+                ? new PollingSubscription(
+                    observer: observer,
+                    pollingInterval: pollingIntervalValue,
+                    scheduler: scheduler,
+                    source: source,
+                    timeSelector: timeSelector)
+                : new OnDemandSubscription(
+                    observer: observer,
+                    scheduler: scheduler,
+                    source: source,
+                    timeSelector: timeSelector));
+        }
+
+        private abstract class SubscriptionBase
+            : IDisposable
+        {
+            private readonly Dictionary<TKey, DateTimeOffset> _expirationDueTimesByKey;
+            private readonly IObserver<IEnumerable<KeyValuePair<TKey, TObject>>> _observer;
+            private readonly List<ProposedExpiration> _proposedExpirationsQueue;
+            private readonly List<KeyValuePair<TKey, TObject>> _removedItemsBuffer;
+            private readonly IScheduler _scheduler;
+            private readonly ISourceCache<TObject, TKey> _source;
+            private readonly IDisposable _sourceSubscription;
+            private readonly Func<TObject, TimeSpan?> _timeSelector;
+
+            private bool _hasSourceCompleted;
+            private ScheduledManagement? _nextScheduledManagement;
+
+            protected SubscriptionBase(
+                IObserver<IEnumerable<KeyValuePair<TKey, TObject>>> observer,
+                IScheduler? scheduler,
+                ISourceCache<TObject, TKey> source,
+                Func<TObject, TimeSpan?> timeSelector)
+            {
+                _observer = observer;
+                _source = source;
+                _timeSelector = timeSelector;
+
+                _scheduler = scheduler ?? GlobalConfig.DefaultScheduler;
+
+                _expirationDueTimesByKey = new();
+                _proposedExpirationsQueue = new();
+                _removedItemsBuffer = new();
+
+                _sourceSubscription = source
+                    .Connect()
+                    // It's important to set this flag outside the context of a lock, because it'll be read outside of lock as well.
+                    .Finally(() => _hasSourceCompleted = true)
+                    .Synchronize(SynchronizationGate)
+                    .SubscribeSafe(Observer.Create<IChangeSet<TObject, TKey>>(
```

Comment:
There's an overload to avoid having to invoke Observer.Create directly.

Reply:
Yeah, I remember looking for it, but I didn't have it on my branch at that time. I should be able to swap them now.

Reply:
Swapped now.